### PR TITLE
STAC Catalog Functionality

### PR DESCRIPTION
**Pull request by GregoryPetrochenkov-NOAA**
This PR addresses #183 as well as bug fixes for attribute tracking and tabulation pairing dictionary attributes.

## Changes

- gval_xarray.py:  Address issue with attribute_tracking names
- tabulationpy:  Address issue with pairing_dictionary attributes in tabulation
- loading_datasets.py: Refactor query STAC function, and implement STAC query to catalog DataFrame
- cases_catalogs.py: Add test cases for STAC catalog comparisons
- test_catalogs.py: Add tests for STAC catalog comparisons
- pyproject.toml:  Update version

### CODE REVIEW

File: src/gval/utils/loading_datasets.py
Comment (fernando-aristizabal) [inline src/gval/utils/loading_datasets.py:1]: Also in `get_stac_data`, why are non-unique timestamps filtered out? Could there be spatial assets across space or search criteria with the same timestamps? And generally speaking what are you seeing with the output of stacstack that requires additional processing?

File: src/gval/utils/loading_datasets.py [hunk 1]
```diff
@@ -439,6 +492,125 @@ def get_stac_data(
...
+            "map_name": map_name,
+            "compare_id": compare_id,
+            "coverage_geometry_type": [stac_item.geometry["type"]] * len_assets,
+            "coverage_geometry_coords": [stac_item.geometry["coordinates"]]
+            * len_assets,
+            "coverage_epsg": ["4326"] * len_assets,
+            "asset_epsg": [stac_item.properties["proj:epsg"]] * len_assets,
+        }
+
+        dfs.append(pd.DataFrame(df_contents))
```
Comment (fernando-aristizabal) [inline src/gval/utils/loading_datasets.py:544]: Appending is a big time suck if this starts getting big. The array's have to get readdressed in memory. Anyway a df can be pre-initialized with the column names then Nones. Every row just writes the row out?

**Review by fernando-aristizabal — COMMENTED** [review]
[Removed]

File: src/gval/utils/loading_datasets.py [hunk 2]
```diff
@@ -323,6 +324,58 @@ def _set_crs(stack: xr.DataArray, band_metadata: list = None) -> Number:
         return stack.rio.write_crs(f"EPSG:{band_metadata['epsg'].values}")
 
 
+def query_stac(
```
[the remark this answers is not in the document]
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:327]: I'll make this private

File: src/gval/utils/loading_datasets.py [hunk 3]
```diff
@@ -323,6 +324,58 @@ def _set_crs(stack: xr.DataArray, band_metadata: list = None) -> Number:
...
+    """
+
+    if not isinstance(collections, Iterable):
+        collections = [collections]
+
+    catalog = pystac_client.Client.open(url)
+
+    return catalog.search(
+        datetime=time,
+        collections=[collections],
```
[the remark this answers is not in the document]
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:371]: I'll change to check for str instead of iterable

[the remark this answers is not in the document]
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:371]: Removed pystac_client from code.

File: src/gval/utils/loading_datasets.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:327]: Removed function from loading_datasets.py

File: src/gval/utils/loading_datasets.py [hunk 4]
```diff
@@ -421,86 +352,88 @@ def get_stac_data(
...
+            bands[0] if isinstance(bands, list) else bands
+        )
 
+    return stack
 
-def _stac_to_df(stac_items: list, assets: list = None) -> pd.DataFrame:
-    """Convert a list of stac items to a DataFrame
+
+def stac_to_df(
+    stac_items: ItemCollection,
```
Comment (fernando-aristizabal) [inline src/gval/utils/loading_datasets.py:420]: Could we allow for multiple types of inputs here? STAC items, collections, and catalogs? Would that add too much complexity?

Comment (fernando-aristizabal) [inline src/gval/utils/loading_datasets.py:420]: Can you update variable names and docstring to the new collection input or whatever it ends up being?

File: src/gval/utils/loading_datasets.py [hunk 5]
```diff
@@ -421,86 +352,88 @@ def get_stac_data(
...
+    assets: list = None,
+    column_allow_list: list = None,
```
Comment (fernando-aristizabal) [inline src/gval/utils/loading_datasets.py:422]: Instead of column, could we say attribute to be more consistent?

**Review by fernando-aristizabal — CHANGES_REQUESTED** [review]

File: src/gval/utils/loading_datasets.py [hunk 6]
```diff
@@ -368,75 +352,154 @@ def get_stac_data(
...
+    Raises
+    ------
+    ValueError
+        No entries in DataFrame due to nonexistent asset
+
+    """
+
+    item_dfs, compare_idx = [], 1
+
+    # Iterate through each STAC Item
```
[the remark this answers is not in the document]
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:452]: Simplified with pd.json_normalize.

File: src/gval/utils/loading_datasets.py (hunk 5 above)
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:422]: Changed name from column_*_list to attribute_*_list

File: src/gval/utils/loading_datasets.py [hunk 7]
```diff
@@ -324,93 +325,23 @@ def _set_crs(stack: xr.DataArray, band_metadata: list = None) -> Number:
...
-        datetime=time,
-        collections=[collections],
-        max_items=max_items,
-        intersects=intersects,
-        bbox=bbox,
-        query=query,
-    ).item_collection()
-
-
 def get_stac_data(
```
[the remark this answers is not in the document]
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:328]: I removed the method to convert a STAC query directly to an Xarray object.

File: tests/cases_catalogs.py [hunk 8]
```diff
@@ -461,3 +461,1155 @@ def case_compare_catalogs_fail(
...
     )
+
+
+url = "https://earth-search.aws.element84.com/v1"
+collection = "sentinel-2-l2a"
+times = ["2020-04-01", "2020-04-03"]
+bbox = [-105.78, 35.79, -105.72, 35.84]
+assets = ["aot"]
+
+expected_stac_df = [
```
Comment (fernando-aristizabal) [inline tests/cases_catalogs.py:472]: Given the size, would it be better to save as its own file? Either in the repo or in the S3?

File: src/gval/utils/loading_datasets.py [hunk 9]
```diff
@@ -238,205 +236,81 @@ def _convert_to_dataset(xr_object=Union[xr.DataArray, xr.Dataset]) -> xr.Dataset
...
-
+    item_dfs, compare_idx = [], 1
 
-def _set_crs(stack: xr.DataArray, band_metadata: list = None) -> Number:
-    """
+    # Iterate through each STAC Item and make a unique row for each asset
+    for item in stac_items:
+        item_dict = item.to_dict()
+        item_df = pd.json_normalize(item_dict)
+        mask = item_df.columns.str.contains("assets.*")
```
Comment (fernando-aristizabal) [inline src/gval/utils/loading_datasets.py:276]: What does this `assets.*` line do?

File: src/gval/utils/loading_datasets.py [hunk 10]
```diff
@@ -238,205 +236,81 @@ def _convert_to_dataset(xr_object=Union[xr.DataArray, xr.Dataset]) -> xr.Dataset
...
-    time_aggregate: str = None,
-    max_items: int = None,
-    intersects: dict = None,
-    bbox: list = None,
-    resolution: int = None,
-    nodata_fill: Number = None,
-) -> xr.Dataset:
-    """
+        if len(dfs) < 1:
+            raise ValueError()
```
Comment (fernando-aristizabal) [inline src/gval/utils/loading_datasets.py:300]: Reason?

File: src/gval/utils/loading_datasets.py [hunk 11]
```diff
@@ -238,205 +236,81 @@ def _convert_to_dataset(xr_object=Union[xr.DataArray, xr.Dataset]) -> xr.Dataset
...
-
-        else:
-            stack = stack.to_dataset(name="band_1")
-            _set_nodata(stack["band_1"], band_metadata, nodata_fill)
-            stack["band_1"] = _set_crs(stack["band_1"])
-            stack["band_1"].attrs["original_name"] = (
-                bands[0] if isinstance(bands, list) else bands
-            )
+    if attribute_block_list is not None:
+        catalog_df = catalog_df.drop(attribute_block_list, axis=1)
```
Comment (fernando-aristizabal) [inline src/gval/utils/loading_datasets.py:311]: What if both lists are None? Mutually exclusive arguments?

File: src/gval/utils/loading_datasets.py [hunk 12]
```diff
@@ -238,205 +236,81 @@ def _convert_to_dataset(xr_object=Union[xr.DataArray, xr.Dataset]) -> xr.Dataset
...
+        # Make a unique row for each asset
+        for key, val in item_dict["assets"].items():
+            if assets is None or key in assets:
+                df = pd.json_normalize(val)
+                df["asset"] = key
+                df["compare_id"] = compare_idx
+                df["map_id"] = val["href"]
+                compare_idx += 1
+                concat_df = pd.concat([og_df, df], axis=1)
+                dfs.append(concat_df.loc[:, ~concat_df.columns.duplicated()])
```
Comment (fernando-aristizabal) [inline src/gval/utils/loading_datasets.py:297]: Is the final df one row per asset?

File: tests/cases_catalogs.py (hunk 8 above)
Comment (GregoryPetrochenkov-NOAA) [reply tests/cases_catalogs.py:472]: Created files on S3

File: src/gval/utils/loading_datasets.py (hunk 9 above)
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:296]: As discussed they remove asset columns so there can be a row for each asset.

File: src/gval/utils/loading_datasets.py (hunk 10 above)
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py]: Applied reason to ValueError.

File: src/gval/utils/loading_datasets.py (hunk 12 above)
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:317]: Yes there is one row per an asset.

File: src/gval/utils/loading_datasets.py (hunk 11 above)
Comment (GregoryPetrochenkov-NOAA) [reply src/gval/utils/loading_datasets.py:334]: If both lists are none it is fine, I added logic to forbid a column being in both lists.

File: tests/test_catalogs.py [hunk 13]
```diff
@@ -219,3 +221,88 @@ def test_compare_catalogs_fail(
             open_kwargs=open_kwargs,
             agreement_map_field=agreement_map_field,
         )
+
+
```
Comment (fernando-aristizabal) [inline tests/test_catalogs.py:225]: Nice to see it come together

**Review by fernando-aristizabal — APPROVED** [review]

**Pull request merged**